metron-analytics/metron-statistics/README.md

```diff
@@ -0,0 +1,346 @@
+exact median.  Secondly, the statistical state for seasonal data should
+be limited to a fixed, trailing window.  We do this by ensuring that the
+MAD state is mergeable and able to be queried from within the Profiler.
+
+### Example
+
+We will create a dummy data stream of gaussian noise to illustrate how
+to use the MAD functionality along with the profiler to tag messages as
+outliers or not.
+
+To do this, we will create a 
+* data generator
+* parser
+* profiler profile
+* enrichment and threat triage
+
+#### Data Generator
+
+We can create a simple python script to generate a stream of gaussian
+noise at the frequency of one message per second as a python script
+which should be saved at `~/rand_gen.py`:
+```
+#!/usr/bin/python
+import random
+import sys
+import time
+def main():
+  mu = float(sys.argv[1])
+  sigma = float(sys.argv[2])
+  freq_s = int(sys.argv[3])
+  while True:
+    print str(random.gauss(mu, sigma))
+    sys.stdout.flush()
+    time.sleep(freq_s)
+
+if __name__ == '__main__':
+  main()
+```
+
+This script will take the following as arguments:
+* The mean of the data generated
+* The standard deviation of the data generated
+* The frequency (in seconds) of the data generated
+
+#### The Parser
+
+We will create a parser that will take the single numbers in and create
+a message with a field called `value` in them using the `CSVParser`.
+
+Add the following file to
+`$METRON_HOME/config/zookeeper/parsers/mad.json`:
+```
+{
+  "parserClassName" : "org.apache.metron.parsers.csv.CSVParser"
+ ,"sensorTopic" : "mad"
+ ,"parserConfig" : {
+    "columns" : {
+      "value_str" : 0
+                }
+                   }
+ ,"fieldTransformations" : [
+    {
+    "transformation" : "STELLAR"
+   ,"output" : [ "value" ]
+   ,"config" : {
+      "value" : "TO_DOUBLE(value_str)"
+               }
+    }
+                           ]
+}
+```
+
+#### Enrichment and Threat Intel
+
+We will set a threat triage level of `10` if a message generates a outlier score of more than 3.5.
+This cutoff will depend on your data and should be adjusted based on the
+assumed underlying distribution.  Note that under the assumptions of
+normality, MAD will act as a robust estimator of the standard deviation, so the cutoff
+should be considered the number of standard deviations away.  For other
+distributions, there are other interpretations which will make sense in
+the context of measuring the "degree different".  See
+http://eurekastatistics.com/using-the-median-absolute-deviation-to-find-outliers/
+for a brief discussion of this.
+
+Create the following in
+`$METRON_HOME/config/zookeeper/enrichments/mad.json`:
+
+```
+{
+  "index": "mad",
+  "batchSize": 1,
+  "enrichment": {
+    "fieldMap": {
+      "stellar" : {
+        "config" : {
+          "parser_score" : "OUTLIER_MAD_SCORE(OUTLIER_MAD_STATE_MERGE(
+PROFILE_GET( 'sketchy_mad', 'global', 10, 'MINUTES') ), value)"
+         ,"is_alert" : "if parser_score > 3.5 then true else is_alert"
```

Comment:
Because the output of MAD might shift over time (if distribution changes), I thought it best to externalize it as a z-score rather than bake it into state capture since the two really should be independent.